service/src/main/java/bio/terra/landingzone/stairway/flight/create/resource/step/CreateAksCostOptimizationDataCollectionRulesStep.java

```diff
@@ -0,0 +1,241 @@
+  }
+
+  @Override
+  protected void createResource(FlightContext context, ArmManagers armManagers) {
+    var landingZoneId =
+        getParameterOrThrow(
+            context.getInputParameters(), LandingZoneFlightMapKeys.LANDING_ZONE_ID, UUID.class);
+    var logAnalyticsWorkspaceId =
+        getParameterOrThrow(
+            context.getWorkingMap(),
+            CreateLogAnalyticsWorkspaceStep.LOG_ANALYTICS_WORKSPACE_ID,
+            String.class);
+    var aksId = getParameterOrThrow(context.getWorkingMap(), CreateAksStep.AKS_ID, String.class);
+
+    var dataCollectionRuleId = createRule(landingZoneId, logAnalyticsWorkspaceId, context);
+    // associate rule with aks resource.
+    createRuleAssociation(aksId, dataCollectionRuleId);
+
+    context
+        .getWorkingMap()
+        .put(AKS_COST_OPTIMIZATION_DATA_COLLECTION_RULE_ID, dataCollectionRuleId);
+    logger.info(RESOURCE_CREATED, getResourceType(), dataCollectionRuleId, getMRGName(context));
+  }
+
+  @Override
+  protected void deleteResource(String resourceId) {
+    // do nothing
+  }
+
+  @Override
+  protected String getResourceType() {
+    return "AksCostOptimizationDataCollectionRule";
+  }
+
+  @Override
+  protected Optional<String> getResourceId(FlightContext context) {
+    return Optional.empty();
+  }
+
+  @Override
+  public List<ResourceNameRequirements> getResourceNameRequirements() {
+    return List.of(
+        new ResourceNameRequirements(
+            getResourceType(), ResourceNameGenerator.MAX_DATA_COLLECTION_RULE_NAME_LENGTH));
+  }
+
+  private String createRule(
+      UUID landingZoneId, String logAnalyticsWorkspaceId, FlightContext context) {
+    final String DESTINATION_NAME = "lz_workspace";
+
+    var dataCollectionRulesName = resourceNameProvider.getName(getResourceType());
+
+    try {
+      var dataCollectionRules =
+          armManagers
+              .monitorManager()
+              .serviceClient()
+              .getDataCollectionRules()
+              .createWithResponse(
+                  getMRGName(context),
+                  dataCollectionRulesName,
+                  new DataCollectionRuleResourceInner()
+                      .withDataSources(
+                          new DataCollectionRuleDataSources()
+                              .withExtensions(
+                                  List.of(
+                                      new ExtensionDataSource()
+                                          .withName("ContainerInsightsExtension")
+                                          .withStreams(
+                                              List.of(
+                                                  KnownExtensionDataSourceStreams
+                                                      .MICROSOFT_INSIGHTS_METRICS))
+                                          .withExtensionName("ContainerInsights")
+                                          .withExtensionSettings(defaultExtensionSettings)))
```

Comment:
Interesting, this method just takes an object and hopes for the best, very pythonish.

Reply:
That's a good suggestion especially because this feature is in preview.

Reply:
Yes. That's inconvenient.